protoc-gen-grpc-gateway/internal/gengateway/template.go

```diff
@@ -727,7 +731,7 @@ func (m response_{{$svc.GetName}}_{{$m.GetName}}_{{$b.Index}}) XXX_ResponseBody(
 var (
 	{{range $m := $svc.Methods}}
 	{{range $b := $m.Bindings}}
-	pattern_{{$svc.GetName}}_{{$m.GetName}}_{{$b.Index}} = runtime.MustPattern(runtime.NewPattern({{$b.PathTmpl.Version}}, {{$b.PathTmpl.OpCodes | printf "%#v"}}, {{$b.PathTmpl.Pool | printf "%#v"}}, {{$b.PathTmpl.Verb | printf "%q"}}))
+	pattern_{{$svc.GetName}}_{{$m.GetName}}_{{$b.Index}} = runtime.MustPattern(runtime.NewPattern({{$b.PathTmpl.Version}}, {{$b.PathTmpl.OpCodes | printf "%#v"}}, {{$b.PathTmpl.Pool | printf "%#v"}}, {{$b.PathTmpl.Verb | printf "%q"}}, {{$IgnoreCase}}))
```

Comment:
As mentioned in [the comment](https://github.com/grpc-ecosystem/grpc-gateway/issues/2091#issuecomment-822673474), this breaks backward compatibility.

Reply:
it defaults to false false, so the behavior is the same.
it does break runtime.NewPattern(...) signature.
I can create a separate one if that's what you mean.

runtime.NewCaseInsensitivePattern(...)?
it's a mouthful, but works for me if it works for you :)

Reply:
ideally, the parameters would be collapsed into an option struct, or use functional options, so that they can be modified without breaking the api in the future...

Reply:
hm, I see what you mean with backward compatibility.
so, for minor releases, the requirement is that we can update the package without regenerating the gw?
that's very restrictive, and I would say that any func used in the generated code should be redesigned to have a functional option, or struct base parameter to reduce breaking changes...

Reply:
what about : 

```
type PatternOption func(*Pattern)

func WithIgnoreCase() PatternOption {
	return func(h *Pattern) {
		h.ignoreCase = true
	}
}

NewPattern(version int, ops []int, pool []string, verb string, ...PatternOption) ) (Pattern, error) { }
```

that would allow the old generated code to work with the new version of the package

Reply:
Unfortunately adding a variadic parameter isn't backwards compatible either: https://stackoverflow.com/questions/55163005/could-adding-variadic-parameters-to-a-function-break-existing-code.

Reply:
alright... then I guess we could settle for adding a "pattern.SetIgnoreCase(bool)" func
or make the property public so it can be set after the constructor in the generated file, or before the call to Match in the muxer.

thoughts?

Reply:
I mean there are options, but if you just keep on saying no without offering me a path to resolve it, it's not very productive :). unless you just don't want this feature, in which case I'd appreciate if you'd say so, so that I can just move on and not waste my time

Reply:
I think it's a valid feature so we want it. But it's an open source project on which we work after-hours and it requires different amount of time needed to spot a breaking change versus digging into a problem and finding a viable solution. I can only ask for a bit more patience :)

Let's consider a solution adding a new method:
1. Code generated with version 1 and library in version 2 (with this new public method); it will work fine as the method simply won't be called.
2. Code generated with version 2 and library in version 1; it will break as the generated code will contain the unknown method. But I don't think we promise such guarantees. One should always use the same or never version with regards to which was used to generate the code. @johanbrandhorst can you confirm, please?

If what I described above is correct, a solution with a new constructor like `NewPatternWithOptions()` should also work. It could result in shorted generated code because options can be calculated at the top and usage would look like this: `runtime.MustPattern(runtime.NewPatternWithOptions(..., patternOptions))`.

Reply:
Adam is correct, option #2 should be fine.

Reply:
Thanks for clarifying @adambabik is there any doc about backward/forward compatibility guarantees that should be observed when contributing, and to be expected when using the library?
Good that there is a path forward. I'll update the PR today.

Is there any other concern about the implementation?
documentation needed as part of the PR?
further testing I should write?

I saw that the example have some integration testing. I can change or add one to use ignore_case, and add some integration tests if needed.

Reply:
You're welcome! Unfortunately I don't think there is any doc explaining that. We are trying to follow industry standards.

> Is there any other concern about the implementation?
No.

> documentation needed as part of the PR?
There is no separate documentation for flags so no need. The flag description should be enough.

> further testing I should write?

> I saw that the example have some integration testing. I can change or add one to use ignore_case, and add some integration tests if needed.

An integration test would be nice. If you want to add a new protobuf file which will cover `ignore_case=true`, take a look at `Makefile`'s `proto` target to see how that can be done. Some flags have their own protobuf and `*.buf.gen.yaml` files.

Reply:
Thank you Adam, I had to fix a few template tests for the previous iteration, so that should be fine.

I'll look into adding a separate protobuf file to illustrate and test ignore_case.